dmriprep/utils/vectors.py

```diff
@@ -0,0 +1,334 @@
+    final_fbvec : str
+        File path of saved final b-vectors if `final` is True.
+    """
+    import os
+    from dipy.io import save_pickle, read_bvals_bvecs
+    from dipy.core.gradients import gradient_table
+
+    # Read b-values and b-vectors from disk
+    # Files can be either ‘.bvals’/’.bvecs’ or ‘.txt’ or ‘.npy’ (containing
+    # arrays stored with the appropriate values).
+    if fbval and fbvec:
+        bvals, bvecs = read_bvals_bvecs(fbval, fbvec)
+    else:
+        raise OSError("Either bvals or bvecs files not found (or rescaling failed)!")
+
+    # Creating the gradient table
+    gtab = gradient_table(bvals, bvecs)
+
+    # Correct b0 threshold
+    gtab.b0_threshold = b0_thr
+
+    # Correct b0 mask
+    gtab_bvals = gtab.bvals.copy()
+    b0_thr_ixs = np.where(gtab_bvals < gtab.b0_threshold)[0]
+    gtab_bvals[b0_thr_ixs] = 0
+    gtab.b0s_mask = gtab_bvals == 0
+
+    # Show info
+    print(gtab.info)
+
+    # Save bval/bvec text file equivalents in the case that these are the final
+    # versions of our gradient vectors
+    if final is True:
+        final_bval_path = "%s%s" % (vector_dir, '/final_bval.bval')
+        final_bvec_path = "%s%s" % (vector_dir, '/final_bvec.bvec')
+        np.savetxt(final_bval_path, bvals.astype('int'), fmt='%d', newline=' ')
+        np.savetxt(final_bvec_path, bvecs.T.astype('float'), fmt='%8f')
+        # Save gradient table to pickle
+        gtab_file = "%s%s" % (vector_dir, "/gtab_final.pkl")
+        save_pickle(gtab_file, gtab)
+    else:
+        def next_path(path_pattern):
+            i = 1
+            while os.path.exists(path_pattern % i):
+                i = i * 2
+            a, b = (i // 2, i)
+            while a + 1 < b:
+                c = (a + b) // 2 # interval midpoint
+                a, b = (c, b) if os.path.exists(path_pattern % c) else (a, c)
+            return path_pattern % b
+        # Save gradient table to pickle
+        gtab_file = next_path(vector_dir + "/gtab_%s.pkl")
+        save_pickle(gtab_file, gtab)
+        final_bval_path = None
+        final_bvec_path = None
+
+    return gtab_file, gtab, final_bval_path, final_bvec_path
+
+
+def rescale_bvec(bvec):
+    """
+    Normalizes b-vectors to be of unit length for the non-zero b-values. If the
+    b-vector row reflect a B0, the vector is untouched.
+
+    Parameters
+    ----------
+    bvec : m x n 2d array
+        Raw b-vectors array.
+
+    Returns
+    -------
+    bvec_rescaled : m x n 2d array
+        Unit-normed b-vectors array.
+    """
+    # Enforce proper dimensions
+    bvec = bvec.T if bvec.shape[0] == 3 else bvec
+
+    # Normalize values not close to norm 1
+    bvec_rescaled = [
+        b / np.linalg.norm(b) if not np.isclose(np.linalg.norm(b), 0) else b
+        for b in bvec
+    ]
+    return bvec_rescaled
+
+
+def rescale_bval(bval):
+    """
+    Normalizes b-val by the square of the vector amplitude. If the
```

Comment:
Is this really what this function does? I don't see the vectors being used.

Reply:
I rewrote this function, as it was unfinished when I first opened this PR.